src/Swarm/Game/Scenario/Objective.hs

```diff
@@ -1,24 +1,72 @@
 {-# LANGUAGE OverloadedStrings #-}
 {-# LANGUAGE TemplateHaskell #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
 
 module Swarm.Game.Scenario.Objective where
 
+import Control.Applicative ((<|>))
 import Control.Lens hiding (from, (<.>))
+import Data.Aeson
+import Data.Set qualified as Set
 import Data.Text (Text)
-import Data.Yaml as Y
 import GHC.Generics (Generic)
+import Swarm.Game.Scenario.Objective.Logic as L
 import Swarm.Language.Pipeline (ProcessedTerm)
+import Swarm.TUI.Model.Achievement.Definitions
 import Swarm.Util (reflow)
 
 ------------------------------------------------------------
 -- Scenario objectives
 ------------------------------------------------------------
 
+data PrerequisiteConfig = PrerequisiteConfig
+  { previewable :: Bool
+  -- ^ Typically, only the currently "active" objectives are
+  -- displayed to the user in the Goals dialog. An objective
+  -- is "active" if all of its prerequisites are met.
+  --
+  -- However, some objectives may be "high-level", in that they may
+  -- explain the broader intention behind potentially multiple
+  -- prerequisites.
+  --
+  -- Set this to option True to display this goal in the "upcoming" section even
+  -- if the objective has currently unmet prerequisites.
+  , logic :: Prerequisite ObjectiveLabel
+  -- ^ Boolean expression the represents the condition dependencies which also
+  -- must have been evaluated to True.
+  -- Note that the achievement of these objective dependencies is
+  -- persistent; once achieved, it still counts even if the "condition"
+  -- might not still hold. The condition is never re-evaluated once True.
+  }
+  deriving (Eq, Show, Generic, ToJSON)
+
+instance FromJSON PrerequisiteConfig where
+  -- Parsing JSON/YAML 'PrerequisiteConfig' has three options:
+  --
+  -- 1. simple String case: `prerequisite: enclose_sheep`
+  -- 2. negated form: `prerequisite: not: have_gear`
+  -- 3. full boolean expressions `prerequisite: logic: and: -...`
+  parseJSON val = preString val <|> preObject val
+   where
+    preObject = withObject "prerequisite" $ \v ->
+      PrerequisiteConfig
+        <$> (v .:? "previewable" .!= False)
+        <*> (pNotId v <|> pLogic v)
+    pNotId v = Not . Id <$> v .: "not"
+    pLogic v = v .: "logic"
+    preString = withText "prerequisite" $ \t ->
+      pure (PrerequisiteConfig False $ Id t)
```

Comment:
@xsebek I'm wondering why we need to handle `Not` specially here?  I would think that if in this alternation `preString val <|> preObject val` we automatically promote a bare string to an `"id"`, then this would apply automatically to an `"id"` that exists within a `"not"` condition.

Perhaps to accomplish this, the special json parsing logic needs to be moved to the `Prerequisite` type instead of the `PrerequisiteConfig` type?

Reply:
Think of it as cutting out the noise from Yaml files. 

Instead of prerequisite.logic.not.id, you get prerequisite.not.

You can also add it to the inner type, to allow prerequisite.logic.not, but that does not seem like a huge win.

Reply:
In the prerequisite: <id> case we are not in logic, so no promotion to Id is going to take place if you use logic.

The not is also not in logic, I wrap both in the constructors to allow this shorthand style.

Reply:
Ok, I've offloaded the logic to parse the `Id`-omission JSON shorthand to the `Prerequisite` type (in `Logic.hs`), which simplifies this `PrerequisiteConfig` parsing a bit.  I'll take the small win :)

I've preserved the shorthand that allows omission of the `"logic"` key.